File: Code/Java/MatrixBenchmark.java
import java.lang.management.ManagementFactory;
import java.lang.management.ThreadMXBean;
import java.util.Random;

public class MatrixBenchmark {

    static final int N = 1024;
    static final int ITER = 10;

    public static double[][] initMatrix(Random rand) {
        double[][] M = new double[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                M[i][j] = rand.nextDouble();
            }
        }
        return M;
    }

    public static double[][] multiplyWithProgress(double[][] A, double[][] B, int iteration) {
        double[][] C = new double[N][N];
        int step = N / 10; // cada 10%
        long startTime = System.currentTimeMillis();

        for (int i = 0; i < N; i++) {

            for (int j = 0; j < N; j++) {
                double sum = 0;
                for (int k = 0; k < N; k++) {
                    sum += A[i][k] * B[k][j];
                }
                C[i][j] = sum;
            }

            // Mostrar progreso: 10%, 20%, ..., 100%
            if ((i + 1) % step == 0) {
                double progress = (i + 1) / (double) N;
                long elapsed = System.currentTimeMillis() - startTime;
                long eta = (long)((elapsed / progress) - elapsed);
                System.out.println("Iteración " + iteration + "/" + ITER +
                        " – " + (int)(progress * 100) + "% completado – ETA: " + (eta / 1000) + "s");
            }
        }
        return C;
    }

    public static void main(String[] args) {

        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
        Random rand = new Random();

        double[] wallTimes = new double[ITER];
        double[] cpuTimes = new double[ITER];
        double[] memoryPeaks = new double[ITER];

        for (int it = 0; it < ITER; it++) {

            double[][] A = initMatrix(rand);
            double[][] B = initMatrix(rand);

            Runtime runtime = Runtime.getRuntime();
            runtime.gc(); // limpiar memoria antes de medir
            long memBefore = (runtime.totalMemory() - runtime.freeMemory()) / 1024;

            long wallStart = System.currentTimeMillis();
            long cpuStart = threadBean.getCurrentThreadCpuTime(); // nanosegundos

            @SuppressWarnings("unused")
            double[][] C = multiplyWithProgress(A, B, it + 1);

            long wallEnd = System.currentTimeMillis();
            long cpuEnd = threadBean.getCurrentThreadCpuTime();

            long memAfter = (runtime.totalMemory() - runtime.freeMemory()) / 1024;

            double wall = (wallEnd - wallStart) / 1000.0;
            double cpu = (cpuEnd - cpuStart) / 1_000_000_000.0; // ns → s
            long memDiff = memAfter - memBefore;

            wallTimes[it] = wall;
            cpuTimes[it] = cpu;
            memoryPeaks[it] = memDiff;

            System.out.println("\n--- Iteración " + (it + 1) + " COMPLETADA ---");
            System.out.printf("Tiempo Wall: %.2f s%n", wall);
            System.out.printf("Tiempo CPU:  %.2f s%n", cpu);
            System.out.println("Memoria inicial: " + memBefore + " KB");
            System.out.println("Memoria final:   " + memAfter + " KB");
            System.out.println("Diferencia:      " + memDiff + " KB\n");
        }

        // Promedios
        double avgWall = 0, avgCpu = 0, avgMem = 0;
        for (int i = 0; i < ITER; i++) {
            avgWall += wallTimes[i];
            avgCpu += cpuTimes[i];
            avgMem += memoryPeaks[i];
        }

        System.out.println("\n=== RESULTADOS PROMEDIO (" + ITER + " iteraciones) ===");
        System.out.printf("Tiempo Wall Promedio: %.2f s%n", avgWall / ITER);
        System.out.printf("Tiempo CPU Promedio:  %.2f s%n", avgCpu / ITER);
        System.out.printf("Diferencia Memoria Promedio: %.2f KB%n", avgMem / ITER);
    }
}

File: Code/Python/MatrixBenchmark.py
import random
import time
import tracemalloc

N = 1024
ITER = 10


def init_matrices():
    A = [[random.random() for _ in range(N)] for _ in range(N)]
    B = [[random.random() for _ in range(N)] for _ in range(N)]
    return A, B


def multiply_with_progress(A, B, iteration):
    C = [[0 for _ in range(N)] for _ in range(N)]
    progress_step = N // 10  # 10% chunks
    start_time = time.time()

    for i in range(N):
        for j in range(N):
            total = 0
            for k in range(N):
                total += A[i][k] * B[k][j]
            C[i][j] = total

        # Progreso cada 10%
        if (i + 1) % progress_step == 0:
            progress = (i + 1) / N
            elapsed = time.time() - start_time
            eta = (elapsed / progress) - elapsed
            print(f"Iteración {iteration}/{ITER} – {int(progress * 100)}% completado – ETA: {int(eta)}s")

    return C


wall_times = []
cpu_times = []
memory_usage = []

for it in range(1, ITER + 1):
    A, B = init_matrices()

    tracemalloc.start()
    mem_before = tracemalloc.get_traced_memory()[0]

    wall_start = time.time()
    cpu_start = time.process_time()

    C = multiply_with_progress(A, B, it)

    wall_end = time.time()
    cpu_end = time.process_time()

    mem_after, mem_peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    wall = wall_end - wall_start
    cpu = cpu_end - cpu_start
    mem_diff = (mem_after - mem_before) / 1024
    mem_peak_kb = mem_peak / 1024

    wall_times.append(wall)
    cpu_times.append(cpu)
    memory_usage.append(mem_peak_kb)

    print(f"\n--- Iteración {it} COMPLETADA ---")
    print(f"Tiempo Wall: {wall:.2f} s")
    print(f"Tiempo CPU:  {cpu:.2f} s")
    print(f"Memoria inicial: {mem_before / 1024:.2f} KB")
    print(f"Memoria final:   {mem_after / 1024:.2f} KB")
    print(f"Uso peak:        {mem_peak_kb:.2f} KB")
    print(f"Diferencia:      {mem_diff:.2f} KB\n")

#Resultados promedio
print(f"\n=== RESULTADOS PROMEDIO ({ITER} iteraciones) ===")
print(f"Tiempo Wall Promedio: {sum(wall_times) / ITER:.2f} s")
print(f"Tiempo CPU Promedio:  {sum(cpu_times) / ITER:.2f} s")
print(f"Memoria Peak Promedio: {sum(memory_usage) / ITER:.2f} KB")